fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java

```diff
@@ -723,6 +724,336 @@ public void updateRecords(@Nonnull Descriptors.FileDescriptor recordsDescriptor)
+        RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = messageType.getOptions()
+                .getExtension(RecordMetaDataOptionsProto.record);
+        if (recordTypeOptions != null && recordTypeOptions.hasUsage()) {
+            return recordTypeOptions.getUsage();
+        }
+        return RecordMetaDataOptionsProto.RecordTypeOptions.Usage.RECORD;
+    }
+
+    private void addFieldToUnion(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileBuilder, @Nonnull DescriptorProtos.DescriptorProto newRecordType) {
+        DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(fileBuilder);
+        DescriptorProtos.FieldDescriptorProto.Builder fieldBuilder = DescriptorProtos.FieldDescriptorProto.newBuilder()
+                .setLabel(DescriptorProtos.FieldDescriptorProto.Label.LABEL_OPTIONAL)
+                .setType(DescriptorProtos.FieldDescriptorProto.Type.TYPE_MESSAGE)
+                .setTypeName(newRecordType.getName())
+                .setName("_" + newRecordType.getName())
+                .setNumber(assignFieldNumber(unionBuilder));
+        unionBuilder.addField(fieldBuilder);
+    }
+
+    private int assignFieldNumber(@Nonnull DescriptorProtos.DescriptorProto.Builder messageType) {
+        if (messageType.getFieldCount() == 0) {
+            return 1;
+        }
+        return messageType.getFieldList().stream().mapToInt(DescriptorProtos.FieldDescriptorProto::getNumber).max().getAsInt() + 1;
+    }
+
+    /**
+     * Deprecate a record type from the meta-data.
+     *
+     * @param recordType the record type to be deprecated
+     */
+    public void deprecateRecordType(@Nonnull String recordType) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_DEPRECATE_RECORD_TYPE, deprecateRecordTypeAsync(recordType));
+    }
+
+    /**
+     * Deprecate a record type from the meta-data asynchronously.
+     *
+     * <p> Note that the record type name must be fully qualified. </p>
+     *
+     * @param recordType the record type to be deprecated
+     * @return a future that completes when the record type is deprecated
+     */
+    @Nonnull
+    public CompletableFuture<Void> deprecateRecordTypeAsync(@Nonnull String recordType) {
+        return loadCurrentProto().thenCompose(metaDataProto -> {
+            RecordMetaDataProto.MetaData.Builder builder = metaDataProto.toBuilder();
+            DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(builder.getRecordsBuilder());
+            if (unionBuilder.getName().equals(recordType)) {
+                throw new MetaDataException("Cannot deprecate the union");
+            }
+
+            // deprecate all fields of type recordType from the union.
+            boolean found = false;
+            for (DescriptorProtos.FieldDescriptorProto.Builder fieldBuilder : unionBuilder.getFieldBuilderList()) {
+                if (fieldBuilder.getTypeName().equals(recordType) && fieldBuilder.hasOptions()) {
```

Comment:
Should this check the type name in its own block and then set the deprecated option inside the block to avoid checking the type name for equality twice? I think then `found = true` can be moved from outside the inner if-block.